### ISSUE TITLE
Ensure nuget based reference assemblies are created for PowerShell

### ISSUE THREAD

**Issue by raghushantha** [PowerShell/PowerShell#1152]
This is for developers using our SDK

**Comment by JamesWTruher** [comment]
The original ones were built in a Razzle environment by taking the original shipping binaries running a couple of tools on them to create reference assemblies. I did this because we wanted to support downlevel releases. If possible, we should create them in our build

**Comment by SteveL-MSFT** [comment]
We should align this with the PowerShell Standard ref assembly work

**Comment by joeyaiello** [comment]
@SteveL-MSFT @lzybkr So this is blocked on #3095? 

If so, this can be moved out to `6.0.0-beta` as we stated DevEx/SDK is the focus for post-beta1.

**Comment by TravisEz13** [comment]
The existing reference package has the following file:
![image](https://cloud.githubusercontent.com/assets/10873629/25866233/9e60db58-34aa-11e7-91be-a0bbefacf8ae.png)

**Comment by lzybkr** [comment]
We definitely don't need most of those - no ISE, no workflow, probably no MMI, and we could easily exclude psreadline.

I think it's worth having multiple nuget packages if we can, one per assembly.

**Comment by TravisEz13** [comment]
@lzybkr I agree with a package per assembly.  I'll start with SMA.  One question that came up is if we should have one package both the reference and actual assembly or separate packages.

**Comment by HemantMahawar** [comment]
I think there are 2 main & different personas 
- Binary module authors - needs only PS Core reference assemblies for compilation
- Application developers - needs actual PS Core runtime to create runspaces etc.

So I would recommend having 2 separate packages even just for SMA. /cc @joeyaiello @SteveL-MSFT

**Comment by SteveL-MSFT** [comment]
That makes sense

**Comment by joeyaiello** [comment]
@TravisEz13 and I just spoke about this. I still feel like we should just continue shipping what we shipped so far for alphas (which is the full assemblies). 

Once we have the "PowerShell Standard" PR closed at #3095, we should make that the reference assembly story, write a blog on developing binary cmdlets on macOS/Linux, etc.

**Comment by daxian-dbw** [comment]
If the reference assembly we are discussing here is the one that will be used as PowerShell Standard framework, then it should be a standard alone NuGet package, because it's an abstraction and there is no specific runtime DLL exactly maps to the API surfaces exposed from it.

However, if the reference assembly in the discussion is specific to PowerShell Core, then we should have only one NuGet package that contains both reference assembly and runtime assemblies so that a developer can reference to only one package for both writing a module and hosting powershell.

**Comment by joeyaiello** [comment]
It's my understanding that a developer can use the runtime assembly as a reference assembly at build time. Is this not the case?

**Comment by daxian-dbw** [comment]
That's the case, but the NuGet package will include different runtime assemblies targeting different platforms, and usually, a NuGet package like that will come with a reference assembly as well.

Take `System.Security.Cryptography.ProtectedData` as an example, here is the content of it:
![image](https://cloud.githubusercontent.com/assets/127450/25875483/95294fe4-34cc-11e7-93b8-90ca46ff4e69.png)

**Comment by daxian-dbw** [comment]
Please see my comment at https://github.com/PowerShell/PowerShell/issues/3417#issuecomment-300648549

In a nutshell, I think we need 2 sets of NuGet packages. Take `System.Management.Automation` as an instance, we need
1. A NuGet package that contains only the __*PowerShell Standard*__ reference assembly.
    - It should be used when building a binary powershell module targeting `netstandard2.0`

2. A NuGet package specific for `netcoreapp2.0`, which contains the reference assembly, runtime assembly for Win and Unix platform (including `libpsl.so`).
    - It should be used when building an application targeting `netcoreapp2.0`, which host powershell core.

**Comment by Jaykul** [comment]
Just for the record -- this is still not fixed in Beta 4. It's *still __not possible__* to create a project which targets the PowerShell SDK and works for .Net Core _and_ full framework, except by _referencing the old packages_. 

I'm having to do nonsense like this:

```xml
  <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp2.0'">
    <PackageReference Include="System.Management.Automation" Version="6.0.0-beta.4" />
  </ItemGroup>
  <ItemGroup Condition="'$(TargetFramework)' == 'net462'">
    <PackageReference Remove="Microsoft.NETCore.App" />
    <PackageReference Include="System.Management.Automation" Version="6.0.0-alpha17" />
  </ItemGroup>
```

**Comment by rkeithhill** [comment]
@Jaykul I get the condition reference to `net462` but why does the other reference `netcoreapp2.0` instead of `netstandard2.0`?  Is it because this project file is for an `exe`?

**Comment by Jaykul** [comment]
Because *POWERSHELL DOES*. If any library references netcoreapp2.0, then **everyone** has to -- it's turtles all the way down.

**Comment by joeyaiello** [comment]
Per [this Twitter conversation](https://twitter.com/terrajobst/status/892060963720974336) we might be able to become fully `netstandard2.0`.

In any case, we've known about this problem for a long time, and it's what PowerShell Standard #4374 is intended to solve. We have to do that either way, but it would be nice if the 6.0 SMA.dll could also be `netstandard2.0`.

**Comment by daxian-dbw** [comment]
It's not fully `netstandard2.0`, restore would fail if the library is used in a runtime that doesn't support RefEmit. BTW, `System.Management.Automation` still depends on `System.Runtime.Loader` which is only available in `netcoreapp2.0`.

**Comment by Jaykul** [comment]
The point, @daxian-dbw is that your library shouldn't have to know who supports RefEmit, RefEmit knows.

Both System.Runtime.Loader and System.Reflection.Emit are available as nuget packages targeting NetStandard.

**Comment by Jaykul** [comment]
So @rkeithhill pointed out [PowerShellStandard.Library](https://www.nuget.org/packages/PowerShellStandard.Library) ...

I noticed it's version `3.0.0-preview-01`, and only seems to support PowerShell API for 3.0  (no Information stream, for instance). Is the plan to have versions of this to support each API version, or did someone decide nothing that's been done since 2012 matters?

**Comment by SteveL-MSFT** [comment]
@Jaykul the current release 3.0.0 aligns with PowerShell 3.0.  It's a starting point.  We plan to have a 5.x version and probably 6.x.

**Comment by powercode** [comment]
Any progress on this?

I run into issues with `IArgumentCompleter` when compiling, and problems loading System.Net.Http on Desktop when running against the `3.0.0-preview-01`.

**Comment by SteveL-MSFT** [comment]
@powercode hoping to have an update on this by next week

**Comment by JonKohler** [comment]
hey @SteveL-MSFT - Any update on this?

**Comment by SteveL-MSFT** [comment]
@JonKohler sorry, had tweeted it, but forgot to update here https://www.nuget.org/packages/PowerShellStandard.Library/5.1.0-preview-01

**Comment by mallochine** [comment]
@SteveL-MSFT sorry if this is obvious -- how to install via the 'nuget' CLI? For some reason Install-Package didn't work but nuget does. E.g. `nuget install Newtonsoft.Json`

**Comment by mallochine** [comment]
oh, nevermind, figured it out. Needed to add -Version or else it doesn't work:

nuget install PowerShellStandard.Library -Version 5.1.0-preview-01

**Comment by mallochine** [comment]
@SteveL-MSFT another problem -- I noticed the reference assembly installed was net standard 2.0, not 4.5. Thus it didn't compile. What is the correct way to install the right framework?

**Comment by mallochine** [comment]
@SteveL-MSFT  so I see that I was supposed to pass in "-sdk:2" flag to 'mcs' compiler to make this work. However, my library uses 'dynamic' language feature, which is not supported in 2.0.

I mean, there must be a way to use the 5.1.0 preview on other release versions, isn't there?

**Comment by Jaykul** [comment]
@mallochine - SteveL said "for now" but it's just to be clear: this isn't really just "for now" -- if you want to write cmdlets that target _PowerShell Core_, you can't use features that aren't in netstandard2.0. In fact, you must compile targeting .NET Standard 2.0 -- and any Windows machines will need at least .NET Framework 4.6.1 (and possibly 4.7, depending on features you use) if you want to be sure the module will work in full desktop "Windows PowerShell" ...

**Comment by mallochine** [comment]
Ok, so turns out actually .NET Standard and .NET Framework at different things. I finally got this to compile.

How do I install .NET Core 2.0 SDK from nuget? Since the library I used was deprecated -- https://www.nuget.org/packages/NETStandard.Library.NETFramework/2.0.0-preview2-25405-01

**Comment by darkonejr** [comment]
Is this truly based on PowerShell 5.1?  Or is this just renamed from the original PowerShell 3 version?  I don't see SMA.Runspace.CreatePipeline() for instance.

**Comment by SteveL-MSFT** [comment]
@darkonejr PowerShell Standard Library intends to always be forward compatible.  As such, PS Std Lib v3 is a subset of PS v3 where the apis still exist in PSCore6.  Similarly , PS Std lib v5.1 is a subset of PS v5.1 where the apis still exist in PSCore6 (essentially, some exceptions).  I would expect CreatePipeline() to still be there.  If its not, it's probably a bug in the preview.  I would suggest opening a new issue for any APIs you expect to be there, but aren't.

**Comment by leonardder** [comment]
Is this meant to be a replacement of Microsoft.PowerShell.5.1.ReferenceAssemblies?
